### PR TITLE
A test that exits early should fail

### PR DESCRIPTION
**Pull request by rictic**
Fixes https://github.com/lukeed/uvu/issues/206

Also fixes the case where a test calls process.exit(0) itself, as that's essentially the same thing.

### CODE REVIEW

**Comment by dkstrong** [comment]
IMO I think it would make more sense to change uvu to not allow the executor to be GCed while it is running. So that the test would more closely match what would happen if you ran the code directly. (e.g. never completes)

**Comment by rolandzwaga** [comment]
I just ran into this issue as well, so regardless of the solution it would be nice to see this fixed.

**Comment by rictic** [comment]
The nice part about this solution is that it also catches the case where you call into a library and that library calls `process.exit(0)`. That library might even be doing a reasonable thing, but we want the test to fail in that case because not all tests may have completed, let alone completed successfully.

This isn't in tension with a solution that prevents tests from being GC'd

File: src/index.js [hunk 1]
```diff
@@ -151,3 +155,12 @@ export async function exec(bail) {
 
 	if (isNode) process.exitCode = code;
 }
+
+process.on('exit', () => {
```
Comment (lukeed) [inline src/index.js:159]: This should be wrapped with a `if (isNode)` condition

File: src/index.js [hunk 2]
```diff
@@ -151,3 +155,12 @@ export async function exec(bail) {
...
+	if (testingInProgress) {
+		console.error(`Exiting early before testing is finished.`);
+		if (process.exitCode === 0 || process.exitCode == null) {
+			process.exitCode = 1;
+		}
```
Comment (lukeed) [inline src/index.js:164]:
```suggestion
		process.exitCode = process.exitCode || 1;
```

File: test/index.js [hunk 3]
```diff
@@ -14,6 +14,22 @@ totalist(__dirname, (rel, abs) => {
 	let pid = spawnSync('node', HOOK.concat(abs));
 	let file = kleur.bold().underline(rel);
 
+	if (rel.endsWith('.fails.js')) {
+		try {
+			assert.notEqual(pid.status, 0, 'expected to fail');
+			assert.equal(pid.stderr.length > 0, true, 'run w/ stderr');
+			assert.equal(pid.stdout.length === 0, true, 'run w/o stdout');
```
Comment (lukeed) [inline test/index.js:21]:
```suggestion
			assert.equal(pid.stdout.length, 0, 'run w/o stdout');
```

**Review by lukeed — COMMENTED** [review]
Looking good! few minor things, thanks!

**Review by lukeed — APPROVED** [review]
Thanks!

**Pull request merged**